Ensembles

PantelisElinas opened the pull request:
Hi @youph 

I have finished the implementation of the Ensemble class. I have also included two Jupyter notebooks demonstrating how to build GraphSAGE, GAT, GCN and HINSAGE ensembles for node attribute inference and link prediction (where applicable.)

I'm copying the reviewer checklist from the original issue #310 here as well.

Regards,

P.

### Reviewer Checklist
- [x] Required functionality implemented
- [x] Unit tests
- [x] Documentation (doc strings, README, Sphinx)
- [x] Python notebooks in demos directory
- [x] Python notebooks run and demonstrate the core ideas

habiba-h commented:
I had a quick look at the 2 notebooks. All runs smoothly and the steps are well explained. 
A few minor things with respect to consistency and some things I am generally curious about:

1. The name of the node classification notebook says Cora but it also has Pubmed data as well. 
2. The title of the Cora link prediction notebook doesn't mention Ensemble (unlike the node classification one) so it isn't clear upfront what the demo is about if one already doesn't know that .
3. Out of curiosity, does the ensemble technique works for other than GraphSAGE? Node classification shows GraphSAGE as well as GCN utility but not the prediction demo. Just curious.
4. Out of curiosity,  in the prediction demo, the Stellargraph edge splitter is used whereas for node classification/attribute inference, sklearn's train_test_split is used, why not the Stellargraph node_splitter?
5. A line about the ensemble technique (for obvious reasons) taking a bit of time to run, would be helpful warning. 
6.  In the node classification notebook, 1st markdown cell: "The Ensemble class allows brings ensemble learning ....", this line needs to be revised. 
7. In link prediction GraphSAGE is referenced but GraphSAGE or GCN are not referenced in classification notebook, it would be good to have the link to the original work in this demo as well.

PantelisElinas commented on .buildkite/pipeline.yml:
```diff
@@ -11,6 +11,8 @@ steps:
           - BUILDKITE_BRANCH
           - AWS_LOGS_BUCKET
+    agents:
+      queue: 't2medium'
```
The unit tests for the Ensemble class require more memory that any other test since we make copies of models. The small instance with 1GB of memory that was running our tests would basically run out of memory. I talked with @TheDen and with some testing we decided that we need a larger instance to run the tests. I've tried to reduce the amount of RAM needed but unfortunately, it was still at the limit and the largest instance is the way to go.

PantelisElinas commented on stellargraph/mapper/node_mappers.py:
```diff
@@ -23,6 +23,7 @@
     "GraphSAGENodeGenerator",
     "HinSAGENodeGenerator",
     "FullBatchNodeGenerator",
+    "FullBatchNodeSequence",
```
It is required by *_generator methods to check the type of generator given is valid.

adocherty replied:
Could we make `FullBatchNodeSequence` a subclass of `NodeSequence`?

PantelisElinas commented:
Hi @youph and @habiba-h 

I have updated the code based on your suggestions. I believe I have addressed all of them.

@habiba-h with regards to your points 3 and 4. We don't have implementation for GCN and GAT link prediction so I cannot add them to the notebook. Also, we don't use node_splitter as it is now deprecated and we should probably removed it soon. We made the decision to use sklearn test_train_split for node attribute inference use case.

Regards,

P.
